File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/RangeMatcher.java
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public record RangeMatcher(int fromCodePoint, int toCodePoint, boolean reversed) implements Matcher {
    public RangeMatcher(int fromCodePoint, int toCodePoint) {
        this(fromCodePoint, toCodePoint, false);
    }

    @Override
    public @Nullable MatchResult matchWith(@NotNull String s, int offset) {
        if (offset >= s.length()) return null;
        final int cp = s.codePointAt(offset);
        return cp >= fromCodePoint && cp <= toCodePoint
                ? new MatchResult(
                        Character.toString(cp),
                        reversed ? toCodePoint - cp + fromCodePoint : cp - fromCodePoint)
                : null;
    }

    @Override
    public int width() {
        return 1 + toCodePoint - fromCodePoint;
    }
}

File: fabric_mod_server/src/main/kotlin/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/RangeMatcher.kt
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern

/**
 * Matches all characters in the inclusive range [fromCodePoint]..[toCodePoint]. The resulting index may be [reversed]
 * @see NamedRangeMatcher
 */
data class RangeMatcher(val fromCodePoint: Int, val toCodePoint: Int, val reversed: Boolean = false) : Matcher {

    override fun matchWith(s: String, offset: Int): MatchResult? {
        if (offset >= s.length) return null
        val cp = s.codePointAt(offset)
        return if (cp in fromCodePoint..toCodePoint) MatchResult(
            Character.toString(cp),
            if (reversed) toCodePoint - cp + fromCodePoint else cp - fromCodePoint
        ) else null
    }

    override val width = 1 + toCodePoint - fromCodePoint
}